lib/linguist/classifier.rb

```diff
@@ -28,40 +31,65 @@ def self.call(blob, possible_languages)
     #
     # db       - Hash classifier database object
     # language - String language of data
-    # data     - String contents of file
+    # data     - String contents of file or array of tokens.
     #
     # Examples
     #
-    #   Classifier.train(db, 'Ruby', "def hello; end")
+    #   Classifier.train!(db, 'Ruby', "def hello; end")
     #
     # Returns nothing.
     #
-    # Set LINGUIST_DEBUG=1 or =2 to see probabilities per-token or
-    # per-language.  See also #dump_all_tokens, below.
+    # Set LINGUIST_DEBUG=1, =2 or =3 to print internal statistics.
     def self.train!(db, language, data)
-      tokens = Tokenizer.tokenize(data)
-
-      db['tokens_total'] ||= 0
-      db['languages_total'] ||= 0
-      db['tokens'] ||= {}
-      db['language_tokens'] ||= {}
-      db['languages'] ||= {}
-
-      tokens.each do |token|
-        db['tokens'][language] ||= {}
-        db['tokens'][language][token] ||= 0
-        db['tokens'][language][token] += 1
-        db['language_tokens'][language] ||= 0
-        db['language_tokens'][language] += 1
-        db['tokens_total'] += 1
-      end
-      db['languages'][language] ||= 0
-      db['languages'][language] += 1
-      db['languages_total'] += 1
+      if data.is_a? String
+        tokens = Tokenizer.tokenize(data)
+      else
+        tokens = data.dup
+      end
+
+      db['vocabulary'] ||= {}
+      # Set hash to autoincremented index value
+      if db['vocabulary'].default_proc.nil?
+        db['vocabulary'].default_proc = proc do |hash, key|
+          hash[key] = hash.length
+        end
+      end
+
+      db['samples'] ||= {}
+      db['samples'][language] ||= []
+
+      termfreq = to_vocabulary_index_termfreq(db['vocabulary'], tokens)
+      db['samples'][language] << termfreq
 
       nil
     end
 
+    # Public: Finalize training.
```

Comment:
Don't you think `graduate!` would be a better name for this method? 😉

Reply:
Makes sense :wink: 

Actually, I'm not so fond of `train!` and `finalize_train!`. I did it this way to minimize API changes, but it may make sense to replace both by a new `train!` (or `fit!`) that takes an enumerable of all traning samples and does the whole thing, without an intermediate state where the classifier object is not valid.